### PR TITLE
Publish new style guide

### PR DESCRIPTION
**Pull request by cgranade**

### CODE REVIEW

File: articles/contributing/style-guide.md [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+In offering the Quantum Development Kit, we strive for function and operation names that help quantum developers write programs that are easy to read and that minimize surprise.
+An important part of that is that when we choose names for functions, operations, and types, we are establishing the *vocabulary* that programmers use to express quantum concepts; with our choices, we either help or hinder them in their effort to clearly communicate.
+This places a responsibility on us to make sure that the names we introduce lend themselves to clarity rather than obscurity.
+In this section, we detail how we meet this obligation in terms of explicit guidance that helps us do the best by the Q# development community.
+
+### Operations and Functions ###
+
+The first thing that a name should establish is whether a given symbol represents a function or an operation.
+The difference between functions and operations is critical to understanding how a block of code behaves.
+To communicate this distinction to programmers, we refer back to that in Q#, quantum computation proceeds through the effects of operations.
```
Comment (cpalmer2020) [inline articles/contributing/style-guide.md]: awkward sentence.

Comment (cgranade) [reply articles/contributing/style-guide.md]: Thanks for the catch! I took a stab at rewording it, please let me know what you think.

File: articles/contributing/style-guide.md [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+| ☑ | `newtype GeneratorTerm` | Use of noun phrase clearly refers to the result of calling the UDT constructor. |
+
+***
+
+### Shorthand and Abbreviations ###
+
+The above advice notwithstanding, there are many forms of shorthand that see common and pervasive use in quantum computing.
+We suggest using existing and common shorthand where it exists, especially for operations that are intrinsic to the operation of a target machine.
+For example, we choose the name `X` instead of `ApplyX`, and `Rz` instead of `RotateAboutZ`.
+When using such shorthand, operation names should be uppercase.
```
Comment (bettinaheim) [inline articles/contributing/style-guide.md]: I would argue they should always be upper case.

File: articles/contributing/style-guide.md [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+
+We suggest:
+
+- Avoid the use of proper nouns in names.
+
+# [Examples](#tab/examples)
+
+***
+
+### Type Conversions ###
```
Comment (bettinaheim) [inline articles/contributing/style-guide.md]: Do you want to write some guideline regarding how to structure libraries? I was thinking something along the lines of conversion functions should be written when … and should go into e.g. a separate file/namespace under the name of ..., and should be aggregated e.g. by user defined type/ compilation unit etc.  Stuff of that sort.

File: articles/contributing/style-guide.md [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+
+# [Guidance](#tab/guidance) 
+
+We suggest:
+
+- For all variable and input names, use `pascalCase` in strong preference to `CamelCase`, `snake_case`, or `ANGRY_CASE`.
+- Input names should be descriptive; avoid one or two letter names where possible.
+- Operations and functions accepting exactly one type argument should denote that type argument by `T` when its role is obvious.
+- If a function or operation takes multiple type arguments, or if the role of a single type argument is not obvious, consider using a short capitalized word prefaced by `T` (e.g.: `TOutput`) for each type.
+- Do not include type names in argument and variable names unless absolutely necessary; this information can and should be provided by your development environment.
```
Comment (bettinaheim) [inline articles/contributing/style-guide.md]: I am lacking imagination here regarding a scenario that qualified as "absolutely necessary"?

File: articles/contributing/style-guide.md [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+  For arrays of qubits in particular, consider denoting such types by `Register` where the name refers to a sequence of qubits that are closely related in some way.
+- Variables used as indices into arrays should begin with `idx` and should be singular (e.g.: `things[idxThing]`).
+  In particular, strongly avoid using single-letter variable names as indices; consider using `idx` at a minimum.
+- Variables used to hold lengths of arrays should begin with `n` and should be pluralized (e.g.: `nThings`).
+
+# [Examples](#tab/examples)
+
+***
+
+## Input Conventions ##
```
Comment (bettinaheim) [inline articles/contributing/style-guide.md]: It may be helpful to also think about and formulate conventions regarding the "aggregation" of arguments (i.e. nesting)?

File: articles/contributing/style-guide.md [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+    body (...) { R(PauliX, theta, qubit); }
+    adjoint (...) { R(PauliX, -theta, qubit); }
+    controlled auto;
+    controlled adjoint auto;
+}
+```
+
+***
+
+## Formatting Conventions ##
```
Comment (bettinaheim) [inline articles/contributing/style-guide.md]: Not sure if you want to mention the formatter here. Granted, it has some limitations related to line breaks and whitespace. And we may also want to wait until we have integrated it better with the editor. It just seems a bit weird to have a Formatting section and not mention the operation to format Q#.

File: articles/contributing/style-guide.md [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+
+# [Examples](#tab/examples)
+
+***
+
+## Naming Conventions ##
+
+In offering the Quantum Development Kit, we strive for function and operation names that help quantum developers write programs that are easy to read and that minimize surprise.
+An important part of that is that when we choose names for functions, operations, and types, we are establishing the *vocabulary* that programmers use to express quantum concepts; with our choices, we either help or hinder them in their effort to clearly communicate.
+This places a responsibility on us to make sure that the names we introduce lend themselves to clarity rather than obscurity.
```
[the remark this answers is not in the document]
Comment (cgranade) [reply articles/contributing/style-guide.md]: It's a slightly esoteric but valid construction, if I understand correctly. I do have an unfortunate tendences towards esoteric constructions, I guess... not that helpful to speak to the importance of clarity in a way that is in and of itself esoteric!

File: articles/contributing/style-guide.md [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+
+|   | Name | Description |
+|---|------|-------------|
+| ☒ | <s>`ApplyDecomposedOperation_`</s> | The underscore `_` should not appear at the end of the name. |
+| ☑ | `_ApplyDecomposedOperation` | The underscore `_` at the beginning clearly indicates that this operation is for internal use only. |
+| ☑ | `ComposeImpl` | The suffix `Impl` clearly indicates that this callable is meant to support the `Compose` callable. |
+
+***
+
+### Variants ###
```
[the remark this answers is not in the document]
Comment (cgranade) [reply articles/contributing/style-guide.md]: Sounds good, I'll look forward to discussing that with you shortly, then.

File: articles/contributing/style-guide.md (hunk 4 above)
Comment (cgranade) [reply articles/contributing/style-guide.md]: I guess I meant for that as a hedge for my own lack of imagination, but I think that's already covered by the initial rule so I'll drop it here.

File: articles/contributing/style-guide.md (hunk 5 above)
Comment (cgranade) [reply articles/contributing/style-guide.md]: Good point. I think I'll defer on that for now, though, and will file it as an issue for future improvement if that's OK?

File: articles/contributing/style-guide.md (hunk 6 above)
Comment (cgranade) [reply articles/contributing/style-guide.md]: My thought was for the style guide to be somewhat independent of editor tooling, but I agree it's relevant here so I'll revise accordingly.

File: articles/contributing/style-guide.md [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+***
+
+## Naming Conventions ##
+
+In offering the Quantum Development Kit, we strive for function and operation names that help quantum developers write programs that are easy to read and that minimize surprise.
+An important part of that is that when we choose names for functions, operations, and types, we are establishing the *vocabulary* that programmers use to express quantum concepts; with our choices, we either help or hinder them in their effort to clearly communicate.
+This places a responsibility on us to make sure that the names we introduce lend themselves to clarity rather than obscurity.
+In this section, we detail how we meet this obligation in terms of explicit guidance that helps us do the best by the Q# development community.
+
+### Operations and Functions ###
```
[the remark this answers is not in the document]
Comment (cgranade) [reply articles/contributing/style-guide.md]: Fair point. I think I can soften a bit to "one of the first things."

File: articles/contributing/style-guide.md [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+ms.author: chgranad
+ms.date: 10/12/2018
+ms.topic: article
+uid: microsoft.quantum.contributing.style
+---
+
+# Q# Style Guide #
+## General Conventions ##
+
+The conventions listed here are suggestions only, and should likely be disregarded when they would result in less readable or useful code.
```
[the remark this answers is not in the document]
Comment (cgranade) [reply articles/contributing/style-guide.md]: Good point, I've modified accordingly.

File: articles/contributing/style-guide.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (cgranade) [reply articles/contributing/style-guide.md]: Modified accordingly.

File: articles/contributing/style-guide.md (hunk 2 above)
Comment (cgranade) [reply articles/contributing/style-guide.md]: Clarified, thanks!

File: articles/contributing/style-guide.md [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+
+
+# [Guidance](#tab/guidance)
+
+We suggest:
+
+- Use verbs for operation names.
+- Use nouns or adjectives for function names.
+- Use nouns for user-defined types.
+- For all callable names, use `CamelCase` in strong preference to `pascalCase`, `snake_case`, or `ANGRY_CASE`.
```
[the remark this answers is not in the document]
Comment (cgranade) [reply articles/contributing/style-guide.md]: Clarified. Thanks for the catch!

File: articles/contributing/style-guide.md (hunk 3 above)
Comment (cgranade) [reply articles/contributing/style-guide.md]: I think that for the most part, that's out of scope for a language style guide. Definitely something we should document somewhere, though. I'll make an issue to add that to the contribution guide.

File: articles/contributing/style-guide.md [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+- `ConstantArray`
+- `Head`
+- `LookupFunction`
+
+In particular, in almost all cases, we suggest using past participles where appropriate to indicate that a function name is strongly connected to an action or side effect elsewhere in a quantum program.
+For example,  `ControlledOnInt` uses the part participle form of the verb "control" to indicate that the function acts as an adjective to modify its argument.
+This name has the additional benefit of matching the semantics of the built-in `Controlled` functor, as discussed further below.
+Similarly, _agent nouns_ can be used to construct function and UDT names from operation names, as in the case of the name `Encoder` for a UDT that is strongly associated with `Encode`.
+
+In common cases, we suggest using particular nouns and adjectives:
```
[the remark this answers is not in the document]
Comment (cgranade) [reply articles/contributing/style-guide.md]: Removed, esp. since much of that is still under active discussion anyway.

File: articles/contributing/style-guide.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (cgranade) [reply articles/contributing/style-guide.md]: If it's alright, I'd like to merge this in for now and update as per our conversation.

File: articles/contributing/style-guide.md (hunk 1 above)
Comment (cpalmer2020) [reply articles/contributing/style-guide.md]: I didn't see a change, so I edited.  Do adjust if needed.

File: articles/contributing/style-guide.md [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+| ☒ | <s>`AsDouble`</s> | The input type is not clear from the function name. |
+| ☒ | <s>`PauliArrFromBoolArr`</s> | The input and output types appear in the wrong order. |
+| ☑ | `ResultArrAsBoolArr` | Both the input types and output types are clear. |
+
+***
+
+### Private or Internal Names ###
+
+> [!NOTE]
+> It is expected that the guidance in this section will change with Q# 1.0.
```
Comment (cpalmer2020) [inline articles/contributing/style-guide.md]: Why is this here?  Can we delete this note?  What is the user supposed to do with this info?

Comment (cgranade) [reply articles/contributing/style-guide.md]: The intent was to provide forewarning to where breaking or incompatable changes are most likely. I'll delete, though.

File: articles/contributing/style-guide.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (bettinaheim) [reply articles/contributing/style-guide.md]: Sure, makes sense.

File: articles/contributing/style-guide.md (hunk 13 above)
Comment (bettinaheim) [reply articles/contributing/style-guide.md]: @cgranade Is there a particular reason that this conversation is still open? 
Other than that, looks good to me, so I'll approve - just make sure to address whatever still needed addressing here.

Comment (cgranade) [reply articles/contributing/style-guide.md]: @bettinaheim: Nope, I just forgot to hit resolve. I'll make sure that any remaining comments are captured as new Issues. Thanks!
